### PR TITLE
feat: use `VarText` datatype

### PR DESCRIPTION
**Pull request by mlutze**
replaces #3418

### CODE REVIEW

**Comment by mlutze** [comment]
Marking as RFR. The next step I think is to make text a required param in Type.freshVar and refactor the calls to it.

File: main/src/ca/uwaterloo/flix/language/fmt/FormatType.scala [hunk 1]
```diff
@@ -257,10 +258,14 @@ object FormatType {
           case Rigidity.Flexible => ""
           case Rigidity.Rigid => "!"
         }
-        val s = prefix + suffix
+        val string = prefix + suffix
         audience match {
-          case Audience.Internal => s
-          case Audience.External => text.getOrElse(s)
+          case Audience.Internal => string
+          case Audience.External => text match { // MATT ? what to do if synthetic?
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/fmt/FormatType.scala:264]: Marking for your attention :)

File: main/src/ca/uwaterloo/flix/language/phase/Typer.scala [hunk 2]
```diff
@@ -779,7 +779,7 @@ object Typer {
 
       case KindedAst.Expression.Scope(sym, exp, evar, loc) =>
         // Introduce a rigid variable for the region of `exp`.
-        val regionVar = Type.freshVar(Kind.Bool, sym.loc, Rigidity.Rigid, Some(sym.text))
+        val regionVar = Type.freshVar(Kind.Bool, sym.loc, Rigidity.Rigid, Ast.VarText.Synthetic(sym.text)) // MATT synthetic correct?
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/phase/Typer.scala:782]: Marking for your attention.

File: main/src/ca/uwaterloo/flix/language/phase/unification/Substitution.scala [hunk 3]
```diff
@@ -34,8 +34,8 @@ object Substitution {
     // Ensure that we do not add any x -> x mappings.
     tpe match {
       case y: Type.Var if x.id == y.sym.id => empty
-      case y: Type.Var if x.text.nonEmpty => Substitution(Map(x -> y.withText(x.text)))
-      case y: Type.Var if y.sym.text.nonEmpty => Substitution(Map(x.withText(y.sym.text) -> y))
+      case y: Type.Var if y.sym.text lessPreciseThan x.text => Substitution(Map(x -> y.withText(x.text)))
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/phase/unification/Substitution.scala:37]: Marking for your attention.

**Review by mlutze — COMMENTED** [review]
@magnus-madsen I added a couple review comments where I'm unsure of the best approach.

File: main/src/ca/uwaterloo/flix/language/phase/Typer.scala (hunk 2 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/Typer.scala:782]: This should not be synthetic.

File: main/src/ca/uwaterloo/flix/language/ast/Ast.scala [hunk 4]
```diff
@@ -537,4 +537,44 @@ object Ast {
...
+      case VarText.Text(_) => 2
+    }
+
+    /**
+      * Returns true if `this` VarText is less precise than `that` VarText.
+      *
+      * More precise text should be preferred when choosing a text to use when substituting.
+      *
+      */
+    def lessPreciseThan(that: VarText): Boolean = this.precision < that.precision
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/Ast.scala:560]: -> isStrictLessPreciseThan

File: main/src/ca/uwaterloo/flix/language/ast/Ast.scala [hunk 5]
```diff
@@ -532,4 +532,44 @@ object Ast {
...
+    /**
+      * Returns true if `this` VarText is less precise than `that` VarText.
+      *
+      * More precise text should be preferred when choosing a text to use when substituting.
+      *
+      */
+    def isStrictlyLessPreciseThan(that: VarText): Boolean = this.precision < that.precision
+  }
+
+  object VarText {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/Ast.scala:558]: Wish we had some better names. Let me think about it for a bit. Other than that, I think we can merge.

File: main/src/ca/uwaterloo/flix/language/fmt/FormatType.scala (hunk 1 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/fmt/FormatType.scala]: Comment can be removed.

File: main/src/ca/uwaterloo/flix/language/ast/Ast.scala [hunk 6]
```diff
@@ -532,4 +532,44 @@ object Ast {
...
+  sealed trait VarText {
+
+    /**
+      * A measure of precision of the text.
+      */
+    private def precision: Int = this match {
+      case VarText.Absent => 0
+      case VarText.Synthetic(_) => 1
+      case VarText.Text(_) => 2
+    }
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/Ast.scala:547]: Named | FakeName | Unnamed?

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/ast/Ast.scala:547]: SourceText | ErsatsText

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/ast/Ast.scala:547]: Label = Text | Hint

Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/ast/Ast.scala:547]: I don't like "label" since it's too easy to ignore.

SourceText | FallbackText ?

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/ast/Ast.scala:547]: Text = Absent | SourceText | FallbackText

And then we want to remove Absent in the next PR? Seems reasonable to me.

**Pull request merged**